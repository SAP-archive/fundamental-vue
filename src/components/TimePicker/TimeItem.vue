<template>
  <div class="fd-time">
    <Time
      :type="showMeridian ? 'hour12' : 'hour24'"
      :value="hour"
      @timeUpdate="updateHour"
      placeholder="hh"
    />

    <Time type="minute" :value="minute" @timeUpdate="updateMinute" placeholder="mm" />

    <Time
      v-if="shouldRenderSeconds"
      type="second"
      :value="second"
      @timeUpdate="updateSecond"
      placeholder="ss"
    />

    <Time
      v-if="showMeridian"
      type="meridian"
      :value="meridian"
      @timeUpdate="updateMeridian"
      placeholder="am"
    />
  </div>
</template>

<script>
import TimeRange from "./mixins/TimeRange";
import Time from "./Time/Time.vue";

export default {
  name: "FdTimeItem",
  mixins: [TimeRange],
  components: { Time },
  props: {
    ariaLabel: { type: String, default: "Time Item" },
    value: String, // validator needs to be added.
    showMeridian: { type: Boolean, default: false }
  },
  computed: {
    shouldRenderSeconds() {
      return this.second != null;
    }
  },
  watch: {
    timeValue: {
      immediate: true,
      handler(newTimeValue) {
        if (newTimeValue != null) {
          const time = newTimeValue.toString().split(" ");
          const timeValue = time[0].split(":");
          [this.hour, this.minute, this.second, this.meridian] = [
            timeValue[0],
            timeValue[1],
            timeValue[2],
            time[1]
          ];
        }
      }
    }
  },
  methods: {
    updateHour(hour) {
      const hourType = this.showMeridian ? "hour12" : "hour24";
      if (
        (Number(this.hour) === Number(this.range[hourType].max) &&
          Number(hour) === Number(this.range[hourType].max) - 1) ||
        (Number(hour) === Number(this.range[hourType].max) &&
          Number(this.hour) === Number(this.range[hourType].max) - 1)
      ) {
        let meridian = this.meridian;
        meridian =
          this.meridian === this.range.meridian.min
            ? this.range.meridian.max
            : this.range.meridian.min;
        this.updateMeridian(meridian);
      }
      this.validateTimeUnit("hour", hour, hourType);
    },
    updateMinute(minute) {
      const minuteValue = this.minute === "00" && minute === "-1" ? this.range.minute.max : minute;
      const updateParameters = {
        timeUnitType: "minute",
        value: minuteValue,
        timeType: "minute",
        upperTimeUnitType: "hour",
        upperTimeUnitUpdateHandler: this.updateHour
      };
      this.updateUnit(updateParameters);
    },
    updateSecond(second) {
      const updateParameters = {
        timeUnitType: "second",
        value: second,
        timeType: "second",
        upperTimeUnitType: "minute",
        upperTimeUnitUpdateHandler: this.updateMinute
      };
      this.updateUnit(updateParameters);
    },
    updateMeridian(meridian) {
      this.meridian = meridian ? meridian : "--";
      this.updateTimeItem();
    },
    updateUnit(updateParameters) {
      let dividend = 0;
      let remainder = 0;
      if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].min) &&
        Number(updateParameters.value) === Number(this.range[updateParameters.timeType].max)
      ) {
        dividend = -1;
        remainder = Number(this.range[updateParameters.timeType].max);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].max) &&
        Number(updateParameters.value) === Number(this.range[updateParameters.timeType].min)
      ) {
        dividend = 1;
        remainder = Number(this.range[updateParameters.timeType].min);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else {
        if (
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            updateParameters.value,
            updateParameters.timeType
          )
        ) {
          remainder =
            Number(updateParameters.value) %
            (Number(this.range[updateParameters.timeType].max) + 1);
          dividend = Math.floor(
            Number(updateParameters.value) / (Number(this.range[updateParameters.timeType].max) + 1)
          );
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            remainder.toString().padStart(2, "0"),
            updateParameters.timeType
          );
        }
      }
      const upperValue = (Number(this[updateParameters.upperTimeUnitType]) + dividend)
        .toString()
        .padStart(2, "0");
      updateParameters.upperTimeUnitUpdateHandler(upperValue);
    },
    validateTimeUnit(unitType, value, type) {
      let valueOutOfRange = false;
      if (this.checkValueRange(value, type)) {
        this[unitType] = value ? value : "--";
      } else if (!isNaN(Number(value))) {
        valueOutOfRange = true;
      } else {
        this[unitType] = "--";
      }
      this.updateTimeItem();
      return valueOutOfRange;
    },
    updateTimeItem() {
      const timeValue = [this.hour, this.minute, this.second].join(":");
      const time = [timeValue, this.meridian].join(" ");
      this.timeValue = time;
      this.$emit("timeItemUpdate", this.timeValue);
    }
  },
  data() {
    return {
      timeValue: this.value,
      hour: "--",
      minute: "--",
      second: "--",
      meridian: "--"
    };
  }
};
</script>
