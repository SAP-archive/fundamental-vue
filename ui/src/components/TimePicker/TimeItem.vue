<template>
  <div class="fd-time">
    <Time
      :type="showMeridian ? 'hour12' : 'hour24'"
      :value="hour"
      @timeUpdate="updateHour"
      placeholder="hh"
    />

    <Time
      type="minute"
      :value="minute"
      @timeUpdate="updateMinute"
      placeholder="mm"
    />

    <Time
      v-if="shouldRenderSeconds"
      type="second"
      :value="second"
      @timeUpdate="updateSecond"
      placeholder="ss"
    />

    <Time
      v-if="showMeridian"
      type="meridian"
      :value="meridian"
      @timeUpdate="updateMeridian"
      placeholder="am"
    />
  </div>
</template>

<script lang="ts">
import { mixins } from "@/mixins";
import TimeRange from "./mixins/TimeRange";
import { TimeType } from "./Time/TimeType";
import Time from "./Time/Time.vue";

const timeUnitMapping = {
  hour: "hour",
  minute: "minute",
  second: "second",
  meridian: "meridian"
};

type TimeUnitType = keyof (typeof timeUnitMapping);
interface TimeUnitUpdateParams {
  timeUnitType: TimeUnitType;
  value: string;
  timeType: TimeType;
  upperTimeUnitType: TimeUnitType;
  upperTimeUnitUpdateHandler: (arg1: string) => void;
}

export default mixins(TimeRange).extend({
  name: "FdTimeItem",
  components: { Time },
  props: {
    ariaLabel: { type: String, default: "Time Item" },
    value: String, // validator needs to be added.
    showMeridian: { type: Boolean, default: false }
  },
  computed: {
    shouldRenderSeconds(): boolean {
      return this.second != null;
    }
  },
  watch: {
    timeValue: {
      immediate: true,
      handler(newTimeValue: string | null) {
        if (newTimeValue != null) {
          const time = newTimeValue.toString().split(" ");
          const timeValue = time[0].split(":");
          [this.hour, this.minute, this.second, this.meridian] = [
            timeValue[0],
            timeValue[1],
            timeValue[2],
            time[1]
          ];
        }
      }
    }
  },
  methods: {
    updateHour(hour: string) {
      const hourType = this.showMeridian ? "hour12" : "hour24";
      if (
        (Number(this.hour) === Number(this.range[hourType].max) &&
          Number(hour) === Number(this.range[hourType].max) - 1) ||
        (Number(hour) === Number(this.range[hourType].max) &&
          Number(this.hour) === Number(this.range[hourType].max) - 1)
      ) {
        let meridian = this.meridian;
        meridian =
          this.meridian === this.range.meridian.min
            ? this.range.meridian.max
            : this.range.meridian.min;
        this.updateMeridian(meridian);
      }
      this.validateTimeUnit("hour", hour, hourType);
    },
    updateMinute(minute: string) {
      const minuteValue =
        this.minute === "00" && minute === "-1"
          ? this.range.minute.max
          : minute;
      const updateParameters: TimeUnitUpdateParams = {
        timeUnitType: "minute",
        value: minuteValue,
        timeType: "minute",
        upperTimeUnitType: "hour",
        upperTimeUnitUpdateHandler: this.updateHour
      };
      this.updateUnit(updateParameters);
    },
    updateSecond(second: string) {
      const updateParameters: TimeUnitUpdateParams = {
        timeUnitType: "second",
        value: second,
        timeType: "second",
        upperTimeUnitType: "minute",
        upperTimeUnitUpdateHandler: this.updateMinute
      };
      this.updateUnit(updateParameters);
    },
    updateMeridian(meridian: string) {
      this.meridian = meridian ? meridian : "--";
      this.updateTimeItem();
    },
    updateUnit(updateParameters: TimeUnitUpdateParams) {
      let dividend = 0;
      let remainder = 0;
      if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].min) &&
        Number(updateParameters.value) ===
          Number(this.range[updateParameters.timeType].max)
      ) {
        dividend = -1;
        remainder = Number(this.range[updateParameters.timeType].max);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else if (
        Number(this[updateParameters.timeUnitType]) ===
          Number(this.range[updateParameters.timeType].max) &&
        Number(updateParameters.value) ===
          Number(this.range[updateParameters.timeType].min)
      ) {
        dividend = 1;
        remainder = Number(this.range[updateParameters.timeType].min);
        this[updateParameters.timeUnitType] = updateParameters.value;
      } else {
        if (
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            updateParameters.value,
            updateParameters.timeType
          )
        ) {
          remainder =
            Number(updateParameters.value) %
            (Number(this.range[updateParameters.timeType].max) + 1);
          dividend = Math.floor(
            Number(updateParameters.value) /
              (Number(this.range[updateParameters.timeType].max) + 1)
          );
          this.validateTimeUnit(
            updateParameters.timeUnitType,
            remainder.toString().padStart(2, "0"),
            updateParameters.timeType
          );
        }
      }
      const upperValue = (
        Number(this[updateParameters.upperTimeUnitType]) + dividend
      )
        .toString()
        .padStart(2, "0");
      updateParameters.upperTimeUnitUpdateHandler(upperValue);
    },
    validateTimeUnit(unitType: TimeUnitType, value: string, type: TimeType) {
      let valueOutOfRange = false;
      if (this.checkValueRange(value, type)) {
        this[unitType] = value ? value : "--";
      } else if (!isNaN(Number(value))) {
        valueOutOfRange = true;
      } else {
        this[unitType] = "--";
      }
      this.updateTimeItem();
      return valueOutOfRange;
    },
    updateTimeItem() {
      const timeValue = [this.hour, this.minute, this.second].join(":");
      const time = [timeValue, this.meridian].join(" ");
      this.timeValue = time;
      this.$emit("timeItemUpdate", this.timeValue);
    }
  },
  data() {
    return {
      timeValue: this.value as string | number | null,
      hour: "--" as string | number,
      minute: "--" as string | number,
      second: "--" as string | number,
      meridian: "--" as string | number
    };
  }
});
</script>
