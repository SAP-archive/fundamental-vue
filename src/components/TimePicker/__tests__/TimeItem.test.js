import { mount } from "@vue/test-utils";
import TimeItem from "../TimeItem.vue";

describe("TimeItem", () => {
  let hour = "11";
  let minute = "05";
  let second = "58";
  let meridian = "am";
  const timeValue = `${hour}:${minute}:${second} ${meridian}`;

  const wrapper = mount(TimeItem, {
    propsData: {
      value: timeValue,
      showMeridian: true
    }
  });

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("has correct data", () => {
    expect(wrapper.vm.$data.hour).toEqual(hour);
    expect(wrapper.vm.$data.minute).toEqual(minute);
    expect(wrapper.vm.$data.second).toEqual(second);
    expect(wrapper.vm.$data.meridian).toEqual(meridian);
    expect(wrapper.vm.$data.timeValue).toEqual(timeValue);
  });

  it("adds/subtracts hours correctly", () => {
    const navUp = wrapper.find(".fd-time__item:nth-child(1) .sap-icon--navigation-up-arrow");
    const navDown = wrapper.find(".fd-time__item:nth-child(1) .sap-icon--navigation-down-arrow");

    navUp.trigger("click");
    expect(wrapper.vm.$data.hour).toEqual("12");

    navUp.trigger("click");
    expect(wrapper.vm.$data.hour).toEqual("01");

    navDown.trigger("click");
    expect(wrapper.vm.$data.hour).toEqual("12");

    navDown.trigger("click");
    expect(wrapper.vm.$data.hour).toEqual("11");
  });

  it("adds/subtracts minutes correctly", () => {
    const navUp = wrapper.find(".fd-time__item:nth-child(2) .sap-icon--navigation-up-arrow");
    const navDown = wrapper.find(".fd-time__item:nth-child(2) .sap-icon--navigation-down-arrow");

    navUp.trigger("click");
    expect(wrapper.vm.$data.minute).toEqual("06");

    navDown.trigger("click");
    expect(wrapper.vm.$data.minute).toEqual("05");
  });

  it("adds/subtracts seconds correctly", () => {
    const navUp = wrapper.find(".fd-time__item:nth-child(3) .sap-icon--navigation-up-arrow");
    const navDown = wrapper.find(".fd-time__item:nth-child(3) .sap-icon--navigation-down-arrow");

    navUp.trigger("click");
    expect(wrapper.vm.$data.second).toEqual("59");

    navUp.trigger("click");
    expect(wrapper.vm.$data.second).toEqual("00");

    navDown.trigger("click");
    expect(wrapper.vm.$data.second).toEqual("59");
  });

  it("flips meridian correctly", () => {
    const navUp = wrapper.find(".fd-time__item:nth-child(4) .sap-icon--navigation-up-arrow");
    const navDown = wrapper.find(".fd-time__item:nth-child(4) .sap-icon--navigation-down-arrow");

    navUp.trigger("click");
    expect(wrapper.vm.$data.meridian).toEqual("pm");

    navUp.trigger("click");
    expect(wrapper.vm.$data.meridian).toEqual("am");

    navDown.trigger("click");
    expect(wrapper.vm.$data.meridian).toEqual("pm");
  });
});
