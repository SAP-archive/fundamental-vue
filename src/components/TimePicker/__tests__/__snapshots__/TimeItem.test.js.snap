// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeItem renders correctly 1`] = `
<div
  class="fd-time"
>
  <div
    aria-label="Time Item"
    class="fd-time__item"
  >
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-up-arrow"
      >
        
    
  
      </button>
    </div>
     
    <div
      class="fd-time__input"
    >
      <input
        class="fd-form__control"
        maxlength="2"
        placeholder="hh"
        type="text"
      />
    </div>
     
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-down-arrow"
      >
        
    
  
      </button>
    </div>
  </div>
   
  <div
    aria-label="Time Item"
    class="fd-time__item"
  >
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-up-arrow"
      >
        
    
  
      </button>
    </div>
     
    <div
      class="fd-time__input"
    >
      <input
        class="fd-form__control"
        maxlength="2"
        placeholder="mm"
        type="text"
      />
    </div>
     
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-down-arrow"
      >
        
    
  
      </button>
    </div>
  </div>
   
  <div
    aria-label="Time Item"
    class="fd-time__item"
  >
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-up-arrow"
      >
        
    
  
      </button>
    </div>
     
    <div
      class="fd-time__input"
    >
      <input
        class="fd-form__control"
        maxlength="2"
        placeholder="ss"
        type="text"
      />
    </div>
     
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-down-arrow"
      >
        
    
  
      </button>
    </div>
  </div>
   
  <div
    aria-label="Time Item"
    class="fd-time__item"
  >
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-up-arrow"
      >
        
    
  
      </button>
    </div>
     
    <div
      class="fd-time__input"
    >
      <input
        class="fd-form__control"
        maxlength="2"
        placeholder="am"
        type="text"
      />
    </div>
     
    <div
      class="fd-time__control"
    >
      <button
        class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-down-arrow"
      >
        
    
  
      </button>
    </div>
  </div>
</div>
`;
